<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.9.2" />
<title>theanine-synthetase.analysis API documentation</title>
<meta name="description" content="Entrypoint analysis script." />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>theanine-synthetase.analysis</code></h1>
</header>
<section id="section-intro">
<p>Entrypoint analysis script.</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">&#34;&#34;&#34;Entrypoint analysis script.&#34;&#34;&#34;

import json
from typing import List, Tuple

from dnds import dnds
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns

from alignment import AlignmentResult, align_sequences
from data_index import CSTSI, CSGSI, CSTSI_PROTEIN, CSGSI_PROTEIN
from dir_utils import get_data, make_output_dir, get_output
from dnds_prep import trim_for_dnds
from options import CLUSTER_COUNTS, DNDS_WINDOW_SIZES
from parse_fasta import parse_fasta
from sliding_window import sliding_window


def make_cluster_graphs(seq_filename: str, alignment_result: AlignmentResult):
    &#34;&#34;&#34;
    Produces graphs with the mismatch cluster sizes and saves them to the
    output directory.
    &#34;&#34;&#34;
    sns.set_theme()

    for clusters in CLUSTER_COUNTS:
        # Build DataFrame of mismatch windows
        df = pd.DataFrame(
            {&#34;clusters&#34;: alignment_result.clustered_mismatches(cluster_count=clusters)}
        )

        sns.relplot(data=df[&#34;clusters&#34;], kind=&#34;line&#34;)

        plt.title(f&#34;Mismatches per alignment cluster ({clusters} clusters)&#34;)
        plt.xlabel(&#34;Cluster #&#34;)
        plt.ylabel(&#34;Mismatches&#34;)

        fig = plt.gcf()
        fig.set_size_inches(7, 8)

        y_max = df[&#34;clusters&#34;].max()
        plt.yticks(np.arange(y_max + 1), np.arange(y_max + 1))
        plt.xticks(np.arange(clusters), np.arange(1, clusters + 1))

        plt.savefig(get_output(f&#34;{seq_filename}_clustered_mismatches_{clusters}.png&#34;))


def make_dnds_graph(
    output_filename: str,
    window_sizes: List[int],
    dnds_ratio_data: List[List[Tuple[int, float]]],
):
    &#34;&#34;&#34;
    Produces a single graph showing dN/dS ratios across a whole sequence and saves
    it to the output directory. `window_sizes` and `dnds_ratio_data` are expected
    to be in the same order with respect to the analyses they represent.
    &#34;&#34;&#34;
    sns.set_theme()

    df = pd.DataFrame(
        {
            f&#34;{window_size}bp&#34;: dnds_ratios
            for window_size, dnds_ratios in zip(window_sizes, dnds_ratio_data)
        }
    )

    sns.relplot(data=df, kind=&#34;line&#34;)

    plt.show()


def sliding_window_dnds(
    sequence_1: str, sequence_2: str, window_size: int
) -&gt; List[Tuple[int, float]]:
    &#34;&#34;&#34;
    Performs a sliding-window dN/dS analysis over the provided sequences.
    Returns a list of tuples (start_base_pair, dnds_ratio).
    &#34;&#34;&#34;
    windows = enumerate(
        zip(
            list(sliding_window(sequence_1, n=window_size)),
            list(sliding_window(sequence_2, n=window_size)),
        )
    )
    return [(i * window_size, dnds(*window_pair)) for i, window_pair in windows]


def analyze(seq1_filename: str, seq2_filename: str, nucleotides: bool = False):
    &#34;&#34;&#34;
    Performs an alignment-based analysis on the sequences in the provided FASTA files.
    `nucleotides` should be `True` if the two filenames refer to nucleotide sequences.
    &#34;&#34;&#34;
    # Read first sequence
    seq_name, seq1 = list(parse_fasta(get_data(seq1_filename)))[0]

    # Analyze second sequence
    print(&#34;Analyzing %s...&#34; % seq2_filename)

    seq2 = list(parse_fasta(get_data(seq2_filename)))[0][1]
    alignment_result = align_sequences(seq2, seq1, nucleotides=nucleotides)

    largest_mismatch_pos, largest_mismatch = alignment_result.largest_mismatch()
    percent_similarity = 1 - (
        alignment_result.hamming_distance() / alignment_result.get_alignment_length()
    )

    for clusters in CLUSTER_COUNTS:
        clustered_mismatches = alignment_result.clustered_mismatches(
            cluster_count=clusters
        )
        clustered_mismatch_variance = alignment_result.clustered_mismatch_variance(
            cluster_count=clusters
        )

        if nucleotides:
            trimmed_alignment_1, trimmed_alignment_2 = trim_for_dnds(alignment_result)
            dnds_ratio_data = [
                sliding_window_dnds(
                    trimmed_alignment_1, trimmed_alignment_2, window_size=i
                )
                for i in DNDS_WINDOW_SIZES
            ]
            make_dnds_graph(
                f&#34;{seq2_filename}_dnds_ratios.png&#34;, DNDS_WINDOW_SIZES, dnds_ratio_data
            )

        # Output Supplementary Data 4
        make_output_dir()
        with open(get_output(f&#34;{seq2_filename}_{clusters}.aln.txt&#34;), &#34;w+&#34;) as f:
            f.write(alignment_result.format_result(line_length=100))

        with open(get_output(f&#34;{seq2_filename}_{clusters}.meta.txt&#34;), &#34;w+&#34;) as f:
            f.write(
                &#34;Formatted metadata -- not for programmatic use.\n\n&#34;
                + f&#34;Information for alignment with {seq_name}:\n\n&#34;
                + f&#34;Percent similarity: {percent_similarity}\n&#34;
                + f&#34;Largest mismatch location: {largest_mismatch_pos}\n&#34;
                + f&#34;Largest mismatch size: {largest_mismatch}bp\n&#34;
                + f&#34;Variance between clusters ({clusters} clusters): {clustered_mismatch_variance}\n&#34;
                + f&#34;Clustered mismatches: {clustered_mismatches}\n&#34;
            )

        with open(get_output(f&#34;{seq2_filename}_{clusters}.meta.json&#34;), &#34;w+&#34;) as f:
            json_output = {
                &#34;percent_similarity&#34;: percent_similarity,
                &#34;largest_mismatch_pos&#34;: largest_mismatch_pos,
                &#34;largest_mismatch&#34;: largest_mismatch,
                &#34;clustered_mismatch_variance&#34;: clustered_mismatch_variance,
                &#34;clustered_mismatches&#34;: clustered_mismatches,
            }

            json.dump(
                json_output,
                f,
            )

    make_cluster_graphs(seq2_filename, alignment_result)


if __name__ == &#34;__main__&#34;:
    analyze(CSTSI, CSGSI, nucleotides=True)
    analyze(CSTSI_PROTEIN, CSGSI_PROTEIN)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="theanine-synthetase.analysis.analyze"><code class="name flex">
<span>def <span class="ident">analyze</span></span>(<span>seq1_filename: str, seq2_filename: str, nucleotides: bool = False)</span>
</code></dt>
<dd>
<div class="desc"><p>Performs an alignment-based analysis on the sequences in the provided FASTA files.
<code>nucleotides</code> should be <code>True</code> if the two filenames refer to nucleotide sequences.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def analyze(seq1_filename: str, seq2_filename: str, nucleotides: bool = False):
    &#34;&#34;&#34;
    Performs an alignment-based analysis on the sequences in the provided FASTA files.
    `nucleotides` should be `True` if the two filenames refer to nucleotide sequences.
    &#34;&#34;&#34;
    # Read first sequence
    seq_name, seq1 = list(parse_fasta(get_data(seq1_filename)))[0]

    # Analyze second sequence
    print(&#34;Analyzing %s...&#34; % seq2_filename)

    seq2 = list(parse_fasta(get_data(seq2_filename)))[0][1]
    alignment_result = align_sequences(seq2, seq1, nucleotides=nucleotides)

    largest_mismatch_pos, largest_mismatch = alignment_result.largest_mismatch()
    percent_similarity = 1 - (
        alignment_result.hamming_distance() / alignment_result.get_alignment_length()
    )

    for clusters in CLUSTER_COUNTS:
        clustered_mismatches = alignment_result.clustered_mismatches(
            cluster_count=clusters
        )
        clustered_mismatch_variance = alignment_result.clustered_mismatch_variance(
            cluster_count=clusters
        )

        if nucleotides:
            trimmed_alignment_1, trimmed_alignment_2 = trim_for_dnds(alignment_result)
            dnds_ratio_data = [
                sliding_window_dnds(
                    trimmed_alignment_1, trimmed_alignment_2, window_size=i
                )
                for i in DNDS_WINDOW_SIZES
            ]
            make_dnds_graph(
                f&#34;{seq2_filename}_dnds_ratios.png&#34;, DNDS_WINDOW_SIZES, dnds_ratio_data
            )

        # Output Supplementary Data 4
        make_output_dir()
        with open(get_output(f&#34;{seq2_filename}_{clusters}.aln.txt&#34;), &#34;w+&#34;) as f:
            f.write(alignment_result.format_result(line_length=100))

        with open(get_output(f&#34;{seq2_filename}_{clusters}.meta.txt&#34;), &#34;w+&#34;) as f:
            f.write(
                &#34;Formatted metadata -- not for programmatic use.\n\n&#34;
                + f&#34;Information for alignment with {seq_name}:\n\n&#34;
                + f&#34;Percent similarity: {percent_similarity}\n&#34;
                + f&#34;Largest mismatch location: {largest_mismatch_pos}\n&#34;
                + f&#34;Largest mismatch size: {largest_mismatch}bp\n&#34;
                + f&#34;Variance between clusters ({clusters} clusters): {clustered_mismatch_variance}\n&#34;
                + f&#34;Clustered mismatches: {clustered_mismatches}\n&#34;
            )

        with open(get_output(f&#34;{seq2_filename}_{clusters}.meta.json&#34;), &#34;w+&#34;) as f:
            json_output = {
                &#34;percent_similarity&#34;: percent_similarity,
                &#34;largest_mismatch_pos&#34;: largest_mismatch_pos,
                &#34;largest_mismatch&#34;: largest_mismatch,
                &#34;clustered_mismatch_variance&#34;: clustered_mismatch_variance,
                &#34;clustered_mismatches&#34;: clustered_mismatches,
            }

            json.dump(
                json_output,
                f,
            )

    make_cluster_graphs(seq2_filename, alignment_result)</code></pre>
</details>
</dd>
<dt id="theanine-synthetase.analysis.make_cluster_graphs"><code class="name flex">
<span>def <span class="ident">make_cluster_graphs</span></span>(<span>seq_filename: str, alignment_result: alignment.AlignmentResult)</span>
</code></dt>
<dd>
<div class="desc"><p>Produces graphs with the mismatch cluster sizes and saves them to the
output directory.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def make_cluster_graphs(seq_filename: str, alignment_result: AlignmentResult):
    &#34;&#34;&#34;
    Produces graphs with the mismatch cluster sizes and saves them to the
    output directory.
    &#34;&#34;&#34;
    sns.set_theme()

    for clusters in CLUSTER_COUNTS:
        # Build DataFrame of mismatch windows
        df = pd.DataFrame(
            {&#34;clusters&#34;: alignment_result.clustered_mismatches(cluster_count=clusters)}
        )

        sns.relplot(data=df[&#34;clusters&#34;], kind=&#34;line&#34;)

        plt.title(f&#34;Mismatches per alignment cluster ({clusters} clusters)&#34;)
        plt.xlabel(&#34;Cluster #&#34;)
        plt.ylabel(&#34;Mismatches&#34;)

        fig = plt.gcf()
        fig.set_size_inches(7, 8)

        y_max = df[&#34;clusters&#34;].max()
        plt.yticks(np.arange(y_max + 1), np.arange(y_max + 1))
        plt.xticks(np.arange(clusters), np.arange(1, clusters + 1))

        plt.savefig(get_output(f&#34;{seq_filename}_clustered_mismatches_{clusters}.png&#34;))</code></pre>
</details>
</dd>
<dt id="theanine-synthetase.analysis.make_dnds_graph"><code class="name flex">
<span>def <span class="ident">make_dnds_graph</span></span>(<span>output_filename: str, window_sizes: List[int], dnds_ratio_data: List[List[Tuple[int, float]]])</span>
</code></dt>
<dd>
<div class="desc"><p>Produces a single graph showing dN/dS ratios across a whole sequence and saves
it to the output directory. <code>window_sizes</code> and <code>dnds_ratio_data</code> are expected
to be in the same order with respect to the analyses they represent.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def make_dnds_graph(
    output_filename: str,
    window_sizes: List[int],
    dnds_ratio_data: List[List[Tuple[int, float]]],
):
    &#34;&#34;&#34;
    Produces a single graph showing dN/dS ratios across a whole sequence and saves
    it to the output directory. `window_sizes` and `dnds_ratio_data` are expected
    to be in the same order with respect to the analyses they represent.
    &#34;&#34;&#34;
    sns.set_theme()

    df = pd.DataFrame(
        {
            f&#34;{window_size}bp&#34;: dnds_ratios
            for window_size, dnds_ratios in zip(window_sizes, dnds_ratio_data)
        }
    )

    sns.relplot(data=df, kind=&#34;line&#34;)

    plt.show()</code></pre>
</details>
</dd>
<dt id="theanine-synthetase.analysis.sliding_window_dnds"><code class="name flex">
<span>def <span class="ident">sliding_window_dnds</span></span>(<span>sequence_1: str, sequence_2: str, window_size: int) ‑> List[Tuple[int, float]]</span>
</code></dt>
<dd>
<div class="desc"><p>Performs a sliding-window dN/dS analysis over the provided sequences.
Returns a list of tuples (start_base_pair, dnds_ratio).</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def sliding_window_dnds(
    sequence_1: str, sequence_2: str, window_size: int
) -&gt; List[Tuple[int, float]]:
    &#34;&#34;&#34;
    Performs a sliding-window dN/dS analysis over the provided sequences.
    Returns a list of tuples (start_base_pair, dnds_ratio).
    &#34;&#34;&#34;
    windows = enumerate(
        zip(
            list(sliding_window(sequence_1, n=window_size)),
            list(sliding_window(sequence_2, n=window_size)),
        )
    )
    return [(i * window_size, dnds(*window_pair)) for i, window_pair in windows]</code></pre>
</details>
</dd>
</dl>
</section>
<section>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="theanine-synthetase" href="index.html">theanine-synthetase</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="theanine-synthetase.analysis.analyze" href="#theanine-synthetase.analysis.analyze">analyze</a></code></li>
<li><code><a title="theanine-synthetase.analysis.make_cluster_graphs" href="#theanine-synthetase.analysis.make_cluster_graphs">make_cluster_graphs</a></code></li>
<li><code><a title="theanine-synthetase.analysis.make_dnds_graph" href="#theanine-synthetase.analysis.make_dnds_graph">make_dnds_graph</a></code></li>
<li><code><a title="theanine-synthetase.analysis.sliding_window_dnds" href="#theanine-synthetase.analysis.sliding_window_dnds">sliding_window_dnds</a></code></li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.9.2</a>.</p>
</footer>
</body>
</html>